storey/flow.py

```diff
@@ -1435,3 +1436,182 @@ def get_table(self, key):
+        block the main event loop thread.
+    * "asyncio" – To run in an asyncio task. This is appropriate for I/O tasks that use asyncio, allowing the event
+        loop to continue running while waiting for a response.
+    * "naive" – To run in the main event loop. This is appropriate only for trivial computation and/or file I/O. It
+        means that the runnable will not actually be run in parallel to anything else.
+
+    Subclasses must also override the run() method with user code that handles the event and returns a result.
+
+    Subclasses may optionally override the init() method if the user's implementation of run() requires prior
+    initialization.
+
+    :param name: Runnable name
+    """
+
+    execution_mechanism = None
+    supported_mechanisms = ("multiprocessing", "threading", "asyncio", "naive")
+
+    # ignore unused keyword arguments such as context which may be passed in by mlrun
+    def __init__(self, name: str, **kwargs):
+        if self.execution_mechanism not in self.supported_mechanisms:
+            raise ValueError(
+                "ParallelExecutionRunnable's execution_mechanism attribute must be overridden with one of: "
+                '"multiprocessing", "threading", "asyncio", "naive"'
+            )
+        self.name = name
+
+    def init(self):
+        """Override this method to add initialization logic."""
+        pass
+
+    def run(self, body, path: str):
+        """
+        Override this method with the code this runnable should run. If execution_mechanism is "asyncio", override
+        run_async() instead.
+
+        :param body: Event body
+        :param path: Event path
+        """
+        return body
+
+    async def run_async(self, body, path: str):
+        """
+        If execution_mechanism is "asyncio", override this method with the code this runnable should run. Otherwise,
+        override run() instead.
+
+        :param body: Event body
+        :param path: Event path
+        """
+        return body
+
+    def _run(self, body, path: str):
+        start = time.monotonic()
+        body = self.run(body, path)
+        end = time.monotonic()
+        return _ParallelExecutionRunnableResult(self.name, body, end - start)
+
+    async def _async_run(self, body, path: str):
+        start = time.monotonic()
+        body = await self.run_async(body, path)
+        end = time.monotonic()
+        return _ParallelExecutionRunnableResult(self.name, body, end - start)
+
+
+class ParallelExecution(Flow):
+    """
+    Runs multiple jobs in parallel for each event.
+
+    :param runnables: A list of ParallelExecutionRunnable instances.
+    :param max_processes: Maximum number of processes to spawn.
+    :param max_threads: Maximum number of threads to start.
+    """
+
+    def __init__(
+        self,
+        runnables: list[ParallelExecutionRunnable],
+        max_processes: Optional[int] = None,
+        max_threads: Optional[int] = None,
```

Comment:
maybe we can add some default

Reply:
Not sure how helpful it would be... It could also be confusing if we want to keep supporting `None` as a way to set no limit.

Reply:
We discussed and decided to add some reasonable limit, but maybe in mlrun and not here (TBD).

Reply:
Actually, there is already a default in `ProcessPoolExecutor` which is [os.cpu_count() or 1](https://github.com/python/cpython/blob/v3.9.21/Lib/concurrent/futures/process.py#L592), and in `ThreadPoolExecutor` which is [min(32, (os.cpu_count() or 1) + 4)](https://github.com/python/cpython/blob/v3.9.21/Lib/concurrent/futures/thread.py#L142).

Although python 3.13 changes these defaults as per https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor and https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor.

Wdyt @davesh0812?

Reply:
I'm adding explicit defaults in anyway.